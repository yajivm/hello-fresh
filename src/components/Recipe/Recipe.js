import React, { useState, useEffect } from "react";
import Card from '../../components/Card';
import { recipes } from "../../mockData/data";
import { useNavigateParams } from "../../utils/helper";
import { DEFAULT_PAGE_NUMBER, ITEMS_COUNT_PER_PAGE } from '../../utils/constant';
import { BackIc, ForwardIc } from "../Icons";
import { GridWrapper, GridContainer, Grid, GreyText, Container, PaginatioNumber, PaginationWrapper, Icon } from './Recipe.styles';

const Recipe = () => {
  const navigate = useNavigateParams();
  const [recipeList, setRecipeList] = useState([]);
  const [pageNumber, setPageNumber] = useState(DEFAULT_PAGE_NUMBER);
  const [showPagination, setShowPagination] = useState(false);
  const [showBackIc, setShowBackIc] = useState(false);
  const [showForwardIc, setShowForwardIc] = useState(!(recipeList.length > ITEMS_COUNT_PER_PAGE));

  const onClickCard = id => {
    navigate(`/recipe-details`, `id=${id}`);
  }

  const deleteCard = id => {
    const filteredData = recipeList.filter(item => item.id !== id);
    setRecipeList(filteredData);
  }

  const editCard = id => {
    console.log('====>', id)
  }

  const filteredItemsForPage = () => {
    let recipesData = [];
    let paginationStaus = false;
    if (recipes.length > ITEMS_COUNT_PER_PAGE) {
      const filteredData = recipes.filter((item, ind) => ind < ITEMS_COUNT_PER_PAGE - 1);
      recipesData = filteredData;
      paginationStaus = true;
    } else {
      recipesData = recipes;
    }
    console.log('=====>', recipesData.length)
    setRecipeList(recipesData);
    setShowPagination(paginationStaus);
  }

  useEffect(() => {
    filteredItemsForPage();
  }, []);

  const numPages = () => {
    return Math.ceil(recipes.length / ITEMS_COUNT_PER_PAGE);
  }

  const onChangePage = page => {
    let recipesData = [];
    if (page < 1) {
      page = 1;
    }
    if (page > numPages()) page = numPages();

    for (var i = (page-1) * ITEMS_COUNT_PER_PAGE; i < (page * ITEMS_COUNT_PER_PAGE) && i < recipes.length; i++) {
      recipesData.push(recipes[i]);
    }
    
    if (page === 1) {
      setShowBackIc(false)
    } else setShowBackIc(true)

    console.log(page === numPages())
    if (page === numPages()) {
      setShowForwardIc(false)
    } else setShowForwardIc(true)

    setRecipeList(recipesData);
    setPageNumber(page);
  }

  const onClickPagination = type => {
    let number = pageNumber;
    switch (type) {
      case 'Back':
        number--;
        break;
      case 'Forward':
        number++;
        break;
      default:
        break;
    }
    onChangePage(number);
  }

  return (
    <Container>
      <GridWrapper>
        <GridContainer>
          {
            (recipeList && recipeList.length > 0) ?
              recipeList.map(({ title, description, img, id }, ind) => (
                <Grid key={`receipie-${ind}`} count={4}>
                  <Card title={title} description={description} imageSrc={img} id={id} onClickCard={onClickCard} showActionIcons onClickDeleteIcon={deleteCard} onClickEditIcon={editCard} />
                </Grid>
              ))
            :
              <GreyText>No Recipe Available</GreyText>
          }
        </GridContainer>
      </GridWrapper>
      {showPagination && (
        <PaginationWrapper>
          <Icon onClick={() => onClickPagination('Back')} isIcHidden={showBackIc}>
            <BackIc />
          </Icon>
          <PaginatioNumber>{pageNumber}</PaginatioNumber>
          <Icon onClick={() => onClickPagination('Forward')} isIcHidden={showForwardIc}>
            <ForwardIc />
          </Icon>
        </PaginationWrapper>
      )}
    </Container>
  )
}

export default Recipe;