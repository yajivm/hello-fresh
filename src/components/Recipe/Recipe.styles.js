import styled from 'styled-components';

export const GridWrapper = styled.div`
  width: 100%;
  overflow: hidden;
`

export const GridContainer = styled.div`
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  justify-content: center;
`

export const Grid = styled.div`
  width: ${({ count }) => count === 4 ? `23.2%` : `31.55%`};
  margin: 0 10px 30px;
  cursor: pointer;

  ${({ count }) => count !== 4 ?
    `
      @media(max-width: 1145px) {
        width: 48%;
      }
      @media(max-width: 1000px) {
        width: 47.5%;
      }
      @media(max-width: 800px) {
        max-width: 100%;
        margin: 0 auto 10px;
        width: 90vw;
      }    
    `
  :
    `@media(max-width: 1110px) {
        width: 47.5%;
      }

      @media(max-width: 800px) {
        max-width: 100%;
        margin: 0 auto 10px;
        width: 90vw;
      }    
    `
  }
`

export const GreyText = styled.p`
  font-size: 40px;
  line-height: 52px;
  text-align: center;
  width: 100%;
  color: rgba(0, 0, 0, 0.2);
  margin-top: 40vh;
`

export const Container = styled.div``