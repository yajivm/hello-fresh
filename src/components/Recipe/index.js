import Recipie from './Recipe';

export default Recipie;