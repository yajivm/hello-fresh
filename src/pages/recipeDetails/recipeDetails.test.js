import React from 'react'
import { render, screen, fireEvent } from '@testing-library/react';
import Recipe from '.'

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => {
  const originalModule = jest.requireActual('react-router-dom');
  return {
    ...originalModule,
    useNavigate: () => mockNavigate,
    useLocation: () => ({ search: '?id=1' }),
  };
});

describe('Recipe Details', () => {
  it('should render component', () => {
    const { container } = render(<Recipe />)

    expect(container).toMatchSnapshot();
  })

  it('should navigate to recipe list page when click the recipe bread crumb', () => {
    render(<Recipe />)

    fireEvent.click(screen.getByTestId('recipe-bread-crumb'))

    expect(mockNavigate).toHaveBeenCalledWith('/recipes-list');
  })
})