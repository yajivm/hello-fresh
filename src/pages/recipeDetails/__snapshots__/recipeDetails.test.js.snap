// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Recipe Details should render component 1`] = `
<div>
  <div
    class="sc-bczRLJ cWPtQe"
  >
    <div
      class="sc-ksZaOG fTVgaS"
    >
      <ul
        class="sc-jSMfEi jhLFCE"
      >
        <li
          class="sc-gKXOVf iWtTFq"
          data-testid="recipe-bread-crumb"
        >
          Recipies
        </li>
        <li
          class="sc-gKXOVf ckmyxU"
        >
          Japanese-Style Beef & Veggies
        </li>
      </ul>
    </div>
    <div
      class="sc-dkzDqf"
    >
      <img
        alt="recipe-banner-image"
        class="sc-hKMtZM gJrAbn"
        data-testid="banner-image"
        height="300"
        src="https://img.hellofresh.com/c_fill,f_auto,fl_lossy,h_432,q_auto,w_768/hellofresh_s3/image/japanese-style-beef-sesame-veggies-cb7e5629.jpg"
      />
    </div>
    <div
      class="sc-gsnTZi ezgmav"
    >
      <div
        class="sc-ftvSup haXJOe"
      >
        <div
          class="sc-papXJ ksiAwV"
        >
          <h2
            class="sc-eCYdqJ jVccvm"
            data-testid="title"
          >
            Japanese-Style Beef & Veggies
          </h2>
          <p
            class="sc-eCYdqJ jVccvm"
            data-testid="description"
          >
            with Rice & Coconut Sweet Chilli Mayo
          </p>
        </div>
        <div
          class="sc-jqUVSM hxUOpr"
        >
          <div
            class="sc-kDDrLX gyyBjO"
          >
            <p
              class="sc-eCYdqJ hCpoza"
              data-testid="longDescription"
            >
              This side dish will be sure to steal the main show tonight! Layer with a base of lemon yoghurt, add dukkah roasted sweet potato and onion and top with fresh mint and flaked almonds for one show-stopping combo.
            </p>
            <div
              class="sc-iBkjds"
            >
              <span
                class="sc-eCYdqJ dkwazc"
                data-testid=""
              >
                Tags:
              </span>
              <ul
                class="sc-jSMfEi kWqKTp"
              >
                <li
                  class="sc-gKXOVf ckmyxU"
                >
                  Not Suitable for Coeliacs
                </li>
                <li
                  class="sc-gKXOVf ckmyxU"
                >
                  Naturally Gluten-Free
                </li>
              </ul>
            </div>
          </div>
          <div
            class="sc-kDDrLX ddxIOQ"
          >
            <ul
              class="sc-jSMfEi gTYGsZ"
            >
              <li
                class="sc-gKXOVf bmioJG"
              >
                Preparation Time
              </li>
              <li
                class="sc-gKXOVf ckmyxU"
              >
                30 minutes
              </li>
            </ul>
            <ul
              class="sc-jSMfEi gTYGsZ"
            >
              <li
                class="sc-gKXOVf bmioJG"
              >
                Cooking difficulty
              </li>
              <li
                class="sc-gKXOVf ckmyxU"
              >
                Easy
              </li>
            </ul>
          </div>
        </div>
      </div>
      <div
        class="sc-jqUVSM hxUOpr"
      >
        <div
          class="sc-kDDrLX gyyBjO"
        >
          <div
            class="sc-ftvSup haXJOe"
          >
            <div
              class="sc-iqcoie"
            >
              <h2
                class="sc-eCYdqJ jVccvm"
                data-testid="ingrediants-title"
                font-size="28"
              >
                Ingredients
              </h2>
              <div
                class="sc-crXcEl jAuCoJ"
              >
                <div
                  class="sc-evZas bdsctm"
                >
                  <p
                    class="sc-eCYdqJ jVccvm"
                    data-testid="ingrediant-name-0-quantity"
                  >
                    3
                  </p>
                  <p
                    class="sc-eCYdqJ jVccvm"
                    data-testid="ingrediant-name-0"
                  >
                    sweet potato
                  </p>
                </div>
                <div
                  class="sc-evZas bdsctm"
                >
                  <p
                    class="sc-eCYdqJ jVccvm"
                    data-testid="ingrediant-name-1-quantity"
                  >
                    1 sachet
                  </p>
                  <p
                    class="sc-eCYdqJ jVccvm"
                    data-testid="ingrediant-name-1"
                  >
                    dukkah
                  </p>
                </div>
                <div
                  class="sc-evZas bdsctm"
                >
                  <p
                    class="sc-eCYdqJ jVccvm"
                    data-testid="ingrediant-name-2-quantity"
                  >
                    1
                  </p>
                  <p
                    class="sc-eCYdqJ jVccvm"
                    data-testid="ingrediant-name-2"
                  >
                    lemon
                  </p>
                </div>
                <div
                  class="sc-evZas bdsctm"
                >
                  <p
                    class="sc-eCYdqJ jVccvm"
                    data-testid="ingrediant-name-3-quantity"
                  >
                    1
                  </p>
                  <p
                    class="sc-eCYdqJ jVccvm"
                    data-testid="ingrediant-name-3"
                  >
                    long red chilli
                  </p>
                </div>
                <div
                  class="sc-evZas bdsctm"
                >
                  <p
                    class="sc-eCYdqJ jVccvm"
                    data-testid="ingrediant-name-4-quantity"
                  >
                    1
                  </p>
                  <p
                    class="sc-eCYdqJ jVccvm"
                    data-testid="ingrediant-name-4"
                  >
                    red onion
                  </p>
                </div>
                <div
                  class="sc-evZas bdsctm"
                >
                  <p
                    class="sc-eCYdqJ jVccvm"
                    data-testid="ingrediant-name-5-quantity"
                  >
                    1 packet
                  </p>
                  <p
                    class="sc-eCYdqJ jVccvm"
                    data-testid="ingrediant-name-5"
                  >
                    flaked almonds
                  </p>
                </div>
                <div
                  class="sc-evZas bdsctm"
                >
                  <p
                    class="sc-eCYdqJ jVccvm"
                    data-testid="ingrediant-name-6-quantity"
                  >
                    1
                  </p>
                  <p
                    class="sc-eCYdqJ jVccvm"
                    data-testid="ingrediant-name-6"
                  >
                    mint
                  </p>
                </div>
                <div
                  class="sc-evZas bdsctm"
                >
                  <p
                    class="sc-eCYdqJ jVccvm"
                    data-testid="ingrediant-name-7-quantity"
                  >
                    1 packet
                  </p>
                  <p
                    class="sc-eCYdqJ jVccvm"
                    data-testid="ingrediant-name-7"
                  >
                    Greek-style yoghurt
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="sc-kDDrLX ddxIOQ"
        >
          <div
            class="sc-ftvSup haXJOe"
          >
            <div
              class="sc-iqcoie"
            >
              <div
                class="sc-iBkjds"
              >
                <h3
                  class="sc-eCYdqJ dkwazc"
                  data-testid=""
                >
                  Nutritional Values
                </h3>
                <ul
                  class="sc-jSMfEi uCzug"
                >
                  <li
                    class="sc-gKXOVf bmioJG"
                  >
                    Energy (kJ)
                  </li>
                  <li
                    class="sc-gKXOVf ckmyxU"
                  >
                    0 kJ
                  </li>
                </ul>
                <ul
                  class="sc-jSMfEi uCzug"
                >
                  <li
                    class="sc-gKXOVf bmioJG"
                  >
                    Energy (kJ)
                  </li>
                  <li
                    class="sc-gKXOVf ckmyxU"
                  >
                    1864 kJ
                  </li>
                </ul>
                <ul
                  class="sc-jSMfEi uCzug"
                >
                  <li
                    class="sc-gKXOVf bmioJG"
                  >
                    Fat
                  </li>
                  <li
                    class="sc-gKXOVf ckmyxU"
                  >
                    16.1 g
                  </li>
                </ul>
                <ul
                  class="sc-jSMfEi uCzug"
                >
                  <li
                    class="sc-gKXOVf bmioJG"
                  >
                    of which saturates
                  </li>
                  <li
                    class="sc-gKXOVf ckmyxU"
                  >
                    3 g
                  </li>
                </ul>
                <ul
                  class="sc-jSMfEi uCzug"
                >
                  <li
                    class="sc-gKXOVf bmioJG"
                  >
                    Carbohydrate
                  </li>
                  <li
                    class="sc-gKXOVf ckmyxU"
                  >
                    55.2 g
                  </li>
                </ul>
                <ul
                  class="sc-jSMfEi uCzug"
                >
                  <li
                    class="sc-gKXOVf bmioJG"
                  >
                    of which sugars
                  </li>
                  <li
                    class="sc-gKXOVf ckmyxU"
                  >
                    29.1 g
                  </li>
                </ul>
                <ul
                  class="sc-jSMfEi uCzug"
                >
                  <li
                    class="sc-gKXOVf bmioJG"
                  >
                    Dietary Fibre
                  </li>
                  <li
                    class="sc-gKXOVf ckmyxU"
                  >
                    0 g
                  </li>
                </ul>
                <ul
                  class="sc-jSMfEi uCzug"
                >
                  <li
                    class="sc-gKXOVf bmioJG"
                  >
                    Protein
                  </li>
                  <li
                    class="sc-gKXOVf ckmyxU"
                  >
                    13.2 g
                  </li>
                </ul>
                <ul
                  class="sc-jSMfEi uCzug"
                >
                  <li
                    class="sc-gKXOVf bmioJG"
                  >
                    Cholesterol
                  </li>
                  <li
                    class="sc-gKXOVf ckmyxU"
                  >
                    0 mg
                  </li>
                </ul>
                <ul
                  class="sc-jSMfEi uCzug"
                >
                  <li
                    class="sc-gKXOVf bmioJG"
                  >
                    Sodium
                  </li>
                  <li
                    class="sc-gKXOVf ckmyxU"
                  >
                    543 mg
                  </li>
                </ul>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="sc-ftvSup haXJOe"
      >
        <div
          class="sc-iBkjds"
        >
          <h3
            class="sc-eCYdqJ dkwazc"
            data-testid=""
            font-size="28"
          >
            Utensils
          </h3>
          <ul
            class="sc-jSMfEi kSeqsy"
          >
            <li
              class="sc-gKXOVf ckmyxU"
            >
              Baking Paper
            </li>
            <li
              class="sc-gKXOVf ckmyxU"
            >
              Baking Tray
            </li>
          </ul>
        </div>
      </div>
      <div
        class="sc-ftvSup haXJOe"
      >
        <div
          class="sc-iBkjds"
        >
          <h3
            class="sc-eCYdqJ dkwazc"
            data-testid=""
            font-size="28"
          >
            Instructions
          </h3>
          <div
            class="sc-jqUVSM QrKyo"
          >
            <div
              class="sc-kDDrLX gUGmjP"
            >
              <img
                alt=""
                class="sc-hKMtZM gJrAbn"
                data-testid=""
                src="https://img.hellofresh.com/c_fill,f_auto,fl_lossy,h_432,q_auto,w_768/hellofresh_s3/image/japanese-style-beef-sesame-veggies-cb7e5629.jpg"
              />
            </div>
            <div
              class="sc-kDDrLX gyyBjO"
            >
              <div
                class="sc-breuTD jIEiBj"
              >
                <p
                  class="sc-eCYdqJ jVccvm"
                  data-testid=""
                >
                  Preheat the oven to 220°C/200°C fan-forced. Cut the sweet potato (unpeeled) into wedges. Cut the red onion into 2cm wedges. Place the sweet potato and onion on an oven tray lined with baking paper. Drizzle with olive oil and season with salt and pepper. Toss to coat and sprinkle with the dukkah. Roast until tender, 25-30 minutes.
                </p>
                <p
                  class="sc-eCYdqJ jVccvm"
                  data-testid=""
                >
                  Tips: 
                  Don’t worry if your onion gets a little charred during cooking. It adds to the flavour!
                </p>
              </div>
            </div>
          </div>
          <div
            class="sc-jqUVSM QrKyo"
          >
            <div
              class="sc-kDDrLX gUGmjP"
            >
              <img
                alt=""
                class="sc-hKMtZM gJrAbn"
                data-testid=""
                src="https://img.hellofresh.com/c_fill,f_auto,fl_lossy,h_432,q_auto,w_768/hellofresh_s3/image/japanese-style-beef-sesame-veggies-cb7e5629.jpg"
              />
            </div>
            <div
              class="sc-kDDrLX gyyBjO"
            >
              <div
                class="sc-breuTD jIEiBj"
              >
                <p
                  class="sc-eCYdqJ jVccvm"
                  data-testid=""
                >
                  While the veggies are roasting, heat a small frying pan over a medium-high heat. Add the flaked almonds and toast until golden, 2-3 minutes. Transfer to a plate and set aside.
                </p>
              </div>
            </div>
          </div>
          <div
            class="sc-jqUVSM QrKyo"
          >
            <div
              class="sc-kDDrLX gUGmjP"
            >
              <img
                alt=""
                class="sc-hKMtZM gJrAbn"
                data-testid=""
                src="https://img.hellofresh.com/c_fill,f_auto,fl_lossy,h_432,q_auto,w_768/hellofresh_s3/image/japanese-style-beef-sesame-veggies-cb7e5629.jpg"
              />
            </div>
            <div
              class="sc-kDDrLX gyyBjO"
            >
              <div
                class="sc-breuTD jIEiBj"
              >
                <p
                  class="sc-eCYdqJ jVccvm"
                  data-testid=""
                >
                  When the veggies have 5 minutes remaining, slice the lemon in half. Pick and roughly chop the mint leaves. Thinly slice the long red chilli (if using). In a small bowl combine the Greek yoghurt, a drizzle of olive oil and a generous squeeze of lemon juice. Season to taste.
                </p>
                <p
                  class="sc-eCYdqJ jVccvm"
                  data-testid=""
                >
                  Tips: 
                  Add more or less lemon juice to taste.
                </p>
              </div>
            </div>
          </div>
          <div
            class="sc-jqUVSM QrKyo"
          >
            <div
              class="sc-kDDrLX gUGmjP"
            >
              <img
                alt=""
                class="sc-hKMtZM gJrAbn"
                data-testid=""
                src="https://img.hellofresh.com/c_fill,f_auto,fl_lossy,h_432,q_auto,w_768/hellofresh_s3/image/japanese-style-beef-sesame-veggies-cb7e5629.jpg"
              />
            </div>
            <div
              class="sc-kDDrLX gyyBjO"
            >
              <div
                class="sc-breuTD jIEiBj"
              >
                <p
                  class="sc-eCYdqJ jVccvm"
                  data-testid=""
                >
                  Spread the lemon yoghurt over a serving plate. Top with the dukkah roasted sweet potato and onion. Drizzle with the honey. Garnish with the flaked almonds, mint and chilli (if using).
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
